### PR TITLE
[4.0][plugin][httpheaders] - array to string conversion

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #26501 .

### Summary of Changes
converted array to string

### Testing Instructions
see #26501

### Expected result

no more warning

### CODE REVIEW

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on f2440d683068b16f40426a6a4bb5a2a7fd1640a6<br /><br />Confirm before Patch:
Warning: hash() expects parameter 2 to be string, array given in \plugins\system\httpheaders\httpheaders.php on line 161
Warning: hash() expects parameter 2 to be string, array given in \plugins\system\httpheaders\httpheaders.php on line 172
After Patch: Warnings has gone.

OT: Which Mode should be use?
Custom / Detect or Automatic?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by SharkyKZ** [comment]
Since #25357 scripts/styles are stored as array. So this isn't right, the hash is being generated based on first script/style only.

**Comment by wilsonge** [comment]
@SharkyKZ is right - we probably need to implode to comma separated rather than doing `array_shift`

**Comment by SharkyKZ** [comment]
Maybe we can use existing MD5 hashes?

**Comment by alikon** [comment]
used implode and the md5 value....

**Comment by SharkyKZ** [comment]
No, not like that. I was thinking we could reuse MD5 hashes but says here it's not recommended https://www.w3.org/TR/SRI/#cryptographic-hash-functions.

**Comment by alikon** [comment]
back to implode

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on c50cddb6fd2ab7cede6ad4d7b9d8dee80aaf1ef1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on c50cddb6fd2ab7cede6ad4d7b9d8dee80aaf1ef1<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by alikon** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by SharkyKZ** [comment]
I don't think this is right. @zero-24 can you take a look at this?

**Comment by SharkyKZ** [comment]
This doesn't work. Each script/style tag needs to have its own hash. And the hash must be based on the content exactly as it appears on on the page. The content from head data does not contain formatting added by styles renderer.

**Comment by richard67** [comment]
I have not tested this item.<br /><br />Setting to not tested until clarifications and reviews finished.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by ChristineWk** [comment]
I have not tested this item.<br /><br />OK, I did also set to "Not tested" now.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by brianteeman** [comment]
Please provide real testing instructions - not just see that the warning has gone

File: plugins/system/httpheaders/httpheaders.php [hunk 1]
```diff
@@ -158,7 +158,7 @@ public function applyHashesToCspRule(): void
 
 			foreach ($inlineScripts as $type => $scriptContent)
 			{
-				$scriptHashes[] = "'sha256-" . base64_encode(hash('sha256', $scriptContent, true)) . "'";
+				$scriptHashes[] = "'sha256-" . base64_encode(hash('sha256', implode(',', $scriptContent), true)) . "'";
 			}
```
Comment (zero-24) [inline plugins/system/httpheaders/httpheaders.php]:
```suggestion
			foreach ($inlineScripts as $type => $scriptArray)
			{
				foreach ($scriptArray as $hash => $scriptContent)
				{
					$scriptHashes[] = "'sha256-" . base64_encode(hash('sha256', $scriptContent, true)) . "'";
				}
			}
```

Comment (zero-24) [inline plugins/system/httpheaders/httpheaders.php:165]: new code
```
			foreach ($inlineScripts as $type => $scriptArray)
			{
				foreach ($scriptArray as $hash => $scriptContent)
				{
					$scriptHashes[] = "'sha256-" . base64_encode(hash('sha256', $scriptContent, true)) . "'";
				}
			}
```

File: plugins/system/httpheaders/httpheaders.php [hunk 2]
```diff
@@ -169,7 +169,7 @@ public function applyHashesToCspRule(): void
 
 			foreach ($inlineStyles as $type => $styleContent)
 			{
-				$styleHashes[] = "'sha256-" . base64_encode(hash('sha256', $styleContent, true)) . "'";
+				$styleHashes[] = "'sha256-" . base64_encode(hash('sha256', implode(',', $styleContent), true)) . "'";
```
Comment (zero-24) [inline plugins/system/httpheaders/httpheaders.php]:
```
			foreach ($inlineStyles as $type => $styleArray)
			{
				foreach ($styleArray as $hash => $styleContent)
				{
					$styleHashes[] = "'sha256-" . base64_encode(hash('sha256', $styleContent, true)) . "'";
				}
			}
```

**Review by zero-24 — CHANGES_REQUESTED** [review]
Please see the inline comments

**Comment by alikon** [comment]
thanks @zero24 & @SharkyKZ

**Comment by PoojaPatil1997** [comment]
I have tested this item :white_check_mark: successfully on 53ff2c4b5b1be01af11e1e5d177b291af307264d<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by crommie** [comment]
I have tested this item :white_check_mark: successfully on 53ff2c4b5b1be01af11e1e5d177b291af307264d<br /><br />As expected: warning without the patch, no warning after applying the patch.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by Quy** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26505">issues.joomla.org/tracker/joomla-cms/26505</a>.</sub>

**Comment by SharkyKZ** [comment]
This hides the notice but the code is still broken.

**Comment by zero-24** [comment]
The code is still broken? Can you explain a bit more? As with https://github.com/joomla/joomla-cms/pull/26505/commits/3ae3bcc2c57e10213a140c351ba9e6aea7ff2413 the actual issue should be fixed or do i miss something?

**Comment by SharkyKZ** [comment]
Since https://github.com/joomla/joomla-cms/pull/25357 rendered scripts/styles have formatting added to them. But the hash is generated from scripts/styles without formatting. So the hashes don't match scripts/styles that appear on the page.

Assuming SRI works on pages with MIME type other than `text/html`, this was actually broken since the beginning because we wrap code in CDATA on such pages: https://github.com/joomla/joomla-cms/blob/56d8a357c2a54656fea36ac7e2ec9f20a715c425/libraries/src/Document/Renderer/Html/HeadRenderer.php#L231

**Comment by zero-24** [comment]
Ah ok I get the point now.

**Pull request closed without merging**